### PR TITLE
add TypeScript definitions

### PR DESCRIPTION
**Pull request by vaishnav98**
originally written by @troywweber7
also submitted to @types : https://github.com/DefinitelyTyped/DefinitelyTyped/pull/26814

### CODE REVIEW

File: src/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+
+// OBJECTS
+export const bone: BoneObj;
+
+// FUNCTIONS - TODO obliterate all "any" types where possible! --- TWW
+// https://github.com/jadonk/bonescript#system
+
+export function getPlatform(callback?: (error: ErrorType, platform: any) => void): any;
+
+export function getEeproms(callback?: (error: ErrorType, eeproms: any) => void): any;
```
Comment (jadonk) [inline src/index.d.ts]: This should probably be deprecated from the public interface.

File: src/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+import * as SerialPort from 'serialport';
+import * as VError from 'verror';
+
+// OBJECTS
+export const bone: BoneObj;
+
+// FUNCTIONS - TODO obliterate all "any" types where possible! --- TWW
+// https://github.com/jadonk/bonescript#system
+
+export function getPlatform(callback?: (error: ErrorType, platform: any) => void): any;
```
Comment (jadonk) [inline src/index.d.ts]: Is there a reason that 'platform' is of type 'any' and not a particular structure?

File: src/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+
+export function getEeproms(callback?: (error: ErrorType, eeproms: any) => void): any;
+
+export function echo(data: any, callback?: (error: ErrorType, data: any) => void): any;
+
+export function readTextFile(filename: string, callback?: (error: ErrorType, data: any) => void): any;
```
Comment (jadonk) [inline src/index.d.ts]: Shouldn't 'data' be of a string type?

File: src/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+export const bone: BoneObj;
+
+// FUNCTIONS - TODO obliterate all "any" types where possible! --- TWW
+// https://github.com/jadonk/bonescript#system
+
+export function getPlatform(callback?: (error: ErrorType, platform: any) => void): any;
+
+export function getEeproms(callback?: (error: ErrorType, eeproms: any) => void): any;
+
+export function echo(data: any, callback?: (error: ErrorType, data: any) => void): any;
```
Comment (jadonk) [inline src/index.d.ts]: Shouldn't 'data' on both the call and callback be of a string type?

File: src/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+
+export function readTextFile(filename: string, callback?: (error: ErrorType, data: any) => void): any;
+
+export function writeTextFile(filename: string, data: any, callback?: (error: ErrorType, data: any) => void): void;
```
Comment (jadonk) [inline src/index.d.ts]: Why does the callback have a 'data' value? I think this should only be returning a status of error or not error.

File: src/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+
+export function writeCModule(filename: string, data: any, callback?: (error: ErrorType, data: any) => void): void;
+
+export function setDate(date: any, callback?: (error: ErrorType, data: any) => void): void;
```
Comment (jadonk) [inline src/index.d.ts]: I don't think this should be returning any 'data', just an error status. The 'date' should be a string type with specific formatting.

File: src/index.d.ts [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+
+export function detachInterrupt(pin: string, callback?: DetachIntCb): void;
+
+export function digitalRead(pin: string, callback?: (err: ErrorType, value: number) => void): any;
+
+export function digitalWrite(pin: string, value: PinStateType, callback?: DigitalWriteCb): void;
+
+export function pinMode(
+    pin: string,
+    direction: any,
```
Comment (jadonk) [inline src/index.d.ts]: string.

File: src/index.d.ts [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+    mux?: any,
+    pullup?: any,
```
Comment (jadonk) [inline src/index.d.ts]: string

File: src/index.d.ts [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+    slew?: any,
```
Comment (jadonk) [inline src/index.d.ts]: string. Several of these should probably should be a string enums if typescript supports that.

File: src/index.d.ts [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+    callback?: PinModeCb
+): void;
+
+export function getPinMode(pin: string, callback?: (err: ErrorType, mode: PinModeObj) => void): any;
+
+export function shiftOut(dataPin: any, clockPin: any, bitOrder: any, val: any, callback?: ErrorCb): void;
```
Comment (jadonk) [inline src/index.d.ts]: dataPin and clockPin should be strings.
val should be a number.
I forget what bitOrder should be.

File: src/index.d.ts [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+
+export interface UartsObj {
+    [i: string]: UartInfo;
+}
+
+export interface i2cObj {
+    [i: string]: i2cInfo;
+}
+
+export interface PinInfo {
```
Comment (jadonk) [inline src/index.d.ts]: I'm not sure if we want to include full details here. They are handy, but I'm not 100% committed to keeping them all here. What ones do you think people actually use?

File: src/index.d.ts [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+    // All
+    name: string;
+    options?: string[];
+    // PWM
+    pwm?: PwmInfo;
+    scale?: number;
+    universalName?: string;
+}
+
+export interface PwmInfo {
```
Comment (jadonk) [inline src/index.d.ts]: Really don't think this should be a public interface. I should probably hide it.

File: src/index.d.ts [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+    chip: string;
+    index: number;
+    module: string;
+    muxmode: number;
+    name: string;
+    path: string;
+    sysfs: number;
+}
+
+export interface UartInfo {
```
Comment (jadonk) [inline src/index.d.ts]: Not really a public interface. Referenced only by string.

File: src/index.d.ts [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+    devicetree?: string;
+    rx?: string;
+    tx?: string;
+}
+
+export interface i2cInfo {
```
Comment (jadonk) [inline src/index.d.ts]: Not really a public interface. Referenced only by string.

File: src/index.d.ts [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+    devicetree?: string;
+    path?: string;
+    scl?: string;
+    sda?: string;
+}
+
+// TYPES - adjusted based on observation / testing
```
Comment (jadonk) [inline src/index.d.ts]: Did these get used up above?

Comment (jadonk) [reply src/index.d.ts]: Looks like this is a reasonable way to do string enums!

Comment (vaishnav98) [reply src/index.d.ts]: Yes, most of them were  used above, i have removed some of the exports which were not necessary, had to export them previously due to this rule: https://github.com/Microsoft/dtslint/blob/master/docs/strict-export-declare-modifiers.md

**Comment by vaishnav98** [comment]
Hi, i have tried to make the suggested changes, can  you have a look at it

**Comment by vaishnav98** [comment]
https://github.com/beagleboard/bonescript/issues/46

**Pull request merged**